libr/core/dmh_glibc.inc.c

```diff
@@ -100,32 +100,231 @@ static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
+
+	double version = 0;
 	RDebugMap *map;
 	RListIter *iter;
+	bool found_glibc_map = false;
+
+	if (core->dbg->glibc_version_resolved) {
+		return true;
+	}
+
+	const char *dbg_glibc_version = r_config_get (core->config, "dbg.glibc.version");
+	if (R_STR_ISEMPTY (dbg_glibc_version)) {
+		dbg_glibc_version = NULL;
+	}
+
+	if (dbg_glibc_version)	{
+		// TODO: use ^ and $ which appear to be broken
+		if (!r_regex_match ("\\d.\\d\\d", "e", dbg_glibc_version)) {
+			R_LOG_WARN ("resolve_glibc_version: Unexpected version format in dbg.glibc.version: %s"
+						" (expected format \"\\d.\\d\\d\")", dbg_glibc_version);
+		} else {
+			version = strtod(dbg_glibc_version, NULL);
+			core->dbg->glibc_version = (int) round ((version * 100));
+			core->dbg->glibc_version_d = version;
+			core->dbg->glibc_version_resolved = true;
+			R_LOG_INFO ("libc version %.2f set from dbg.glibc.version", core->dbg->glibc_version_d);
+			return true;
+		}
+	}
+
+	r_return_val_if_fail (core && core->dbg && core->dbg->maps, false);
 	r_debug_map_sync (core->dbg);
+
+	// Search for binary in memory maps named *libc-* or *libc.*
+	// TODO: This is very brittle, other bin names or LD_PRELOAD could be a problem
 	r_list_foreach (core->dbg->maps, iter, map) {
-		// In case the binary is named *libc-* or *libc.*
-		if (strncmp (map->name, core->bin->file, strlen (map->name)) != 0) {
-			char *fp = strstr (map->name, "libc-");
-			if (fp) {
-				break;
-			}
-			fp = strstr (map->name, "libc.");
-			if (fp) {
-				v = r_num_get_double (core->num, fp + 5);
-				core->dbg->glibc_version = (int) round((v * 100));
-				return (v > 2.25);
-			}
+		if (strncmp (map->name, core->bin->file, strlen (map->name)) == 0) {
+			continue;
+		}
+		if (r_regex_match (".*libc[.-]", "e", map->name)) {
+			found_glibc_map = true;
+			break;
 		}
 	}
+	// TODO: handle static binaries
+	if (!found_glibc_map) {
+		R_LOG_WARN ("resolve_glibc_version: no libc found in memory maps (cannot handle static binaries)");
+		return false;
+	}
+	// At this point we found a map in memory that _should_ be libc
+	version = GH (get_glibc_version) (core, map->file);
+	if (version != 0)	{
+		core->dbg->glibc_version = (int) round ((version * 100));
+		core->dbg->glibc_version_d = version;
+		core->dbg->glibc_version_resolved = true;
+		char version_buffer[315] = {0};
+		// TODO: better way snprintf to 4 chars warns
+		// note: ‘snprintf’ output between 4 and 314 bytes into a destination of size 4
+		snprintf (version_buffer, sizeof (version_buffer)-1, "%.2f", version);
+		r_config_set (core->config, "dbg.glibc.version", version_buffer);
+		return true;
+	}
+
 	return false;
 }
 
+static bool GH(is_tcache)(RCore *core) {
+	if (!r_config_get_b (core->config, "cfg.debug")) {
+		return r_config_get_b (core->config, "dbg.glibc.tcache");
+	}
+
+	if (core->dbg->glibc_version_resolved || GH (resolve_glibc_version) (core))	{
+		return core->dbg->glibc_version_d > 2.25;
+	} else {
```

Comment:
no need to else after return

Reply:
It's not guaranteed that libc version can be resolved, that's why the else has a warning printed.